python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -766,13 +785,15 @@ def summary(self):
                 ]
                 is_pending = status in pending_states
                 if is_pending:
-                    pending_nodes.append((ip, node_type, status))
+                    path = self.node_tracker.get_log_path(node_id)
+                    pending_nodes.append((ip, node_type, status, path))
                 else:
-                    # TODO (Alex): Failed nodes are now immediately killed, so
-                    # this list will almost always be empty. We should ideally
-                    # keep a cache of recently failed nodes and their startup
-                    # logs.
-                    failed_nodes.append((ip, node_type))
+                    failed = True
+
+            if not failed:
+                non_failed.add(node_id)
+
+        failed_nodes = self.node_tracker.get_all_failed_node_info(non_failed)
```

Comment:
why this takes as an input non_failed instead of taking a list of failed nodes and get_all_failed_node_info will basically operate on that list?

Reply:
Because the node provider (and thus autoscaler) is not a source of truth about failed nodes